docs/architecture/errors.md

```diff
@@ -0,0 +1,209 @@
+
+ChefRun::Error exposes attributes to control how the error is rendered. See "Rendering" below.
+
+By convention, class-specific errors are defined inline with the same class that raises them.
+
+#### Inline Handling
+In several places we perform inline handling in order to map the error to a chef-run
+error, so that we get proper rendering of the error.
+
+An example of this can be found in `TargetHost#connect!` - it captures a Train::UserError and
+raises a ConnectionFailure with error code and message based on the details it extracted from the
+UserError.
+
+We similarly parse out specific cookbook-related errors during a remote run failure and
+raise an appropriate ChefRun error that includes likely cause/remediation for the erorr.
+
+#### Formatting and Rendering
+This is handled by UI::ErrorPrinter.
+
+The primary interface is `ErrorPrinter.show_error` which accepts a WrappedException and renders it to the terminal
+via `ErrorPrinter#format_error`.
+
+These look at the attributes of the exception (if it's ChefRun::Error-derived) to determine how to format it.
+It renders the error in parts:
+- header: the error ID, in bold
+- body - the customer-facing text of the error
+  - Note that this contains additional handling to do formatting specific to
+   certain error types (chef-run, train, and anything else).  This is an additional
+   place that we do error mapping, in addition to StandardErrorResolver mentioned above.
+  - chef-run errors
+    - have their customer-facing text defined in i18n under the key "errors.ERROR_ID"
+  - train error
+    - these get manually mapped to CHEFTRN001/2 (depending if a TargetHost is available)
+    - the error message is presented as-is to the operator
+  - other errors
+    - these get rendered as CHEFINT001
+    - the error message is presented as-is to the operator
+- footer - the trailing bits of the error
+  - unless suppressed by ChefJob::Error options, we will render a footer using the
+    following i18n keys:
+    - `error.footer.both` - footer gives location of log file and stack trace file
+      - typically used for errors we don't expect to see
+      - maps to ChefRun::Error#show_stack == true && show_log == true
+    - `error.footer.log_only` - footer gives location of log file
+      - maps to ChefRun::Error#show_stack == false && show_log == true
+    - `error.footer.stack_only` - footer gives location of stack trace capture
+      - maps to ChefRun::Error#show_stack == true && show_log == false
+    - `error.footer.neither` - generic message that just says to contact beta@chef.io if you can't fix the error.
+      - used for errors when the cause and remediation are clear, for example
+        bad option arguments to the CLI.
+      - maps to ChefRun::Error#show_stack == false && show_log == false
+
+#### One-off Formatting
+
+The Startup class runs outside of the main error handling, and does its own formatting
+of the errors it expects to see.
+
+
```

Comment:
I think I would like to see
```
#### Remote Chef Client Errors

When `chef-client` fails on the remote node we use a custom error handler to write run
information into a json file on the remote node. We fetch that JSON file and parse it locally on
the workstation to determine the cause of the error. Currently we send back the error text and
use a regular-expression based parser to map remote errors into informative local errors. In
the future we would like to get rid of this regex matching and instead send better information
in the formatted JSON.
```
or something like that

Reply:
I missed this feedback when you posted it - thanks @tball 

I'll update with your suggestions.  The intent here was just to organize thoughts and make a plan to move forward -  not create a living reference to the arch.  Maybe I'll add something to the summary explaining that and a 'true as of mm/dd/yyyy' disclaimer.